boot/freeldr/freeldr/arch/uefi/uefidisk.c

```diff
@@ -0,0 +1,589 @@
+   }
+   else
+   {
+        ULONG BootPartition;
+        PARTITION_TABLE_ENTRY PartitionEntry;
+
+        /* This is a hard disk */
+        if (!UefiGetBootPartitionEntry(FrldrBootDrive, &PartitionEntry, &BootPartition))
+        {
+            ERR("Failed to get boot partition entry\n");
+            return FALSE;
+        }
+
+        RtlStringCbPrintfA(FrLdrBootPath, sizeof(FrLdrBootPath),
+                           "multi(0)disk(0)rdisk(%u)partition(%lu)",
+                           PublicBootArcDisk, BootPartition);
+    }
+
+    return TRUE;
+}
+
+BOOLEAN
+UefiInitializeBootDevices(VOID)
+{
+	ULONG i = 0;
+    DiskReadBufferSize = EFI_PAGE_SIZE;
+    DiskReadBuffer = MmAllocateMemoryWithType(DiskReadBufferSize, LoaderFirmwareTemporary);
+    UefiSetupBlockDevices();
+    UefiSetBootpath();
+
+    /* Add it, if it's a cdrom */
+    GlobalSystemTable->BootServices->HandleProtocol(handles[UefiBootRootIdentifier], &bioGuid, (void **) &bio);
+    if (bio->Media->RemovableMedia == TRUE && bio->Media->BlockSize == 2048)
+    {
+        PMASTER_BOOT_RECORD Mbr;
+        PULONG Buffer;
+        ULONG Checksum = 0;
+        ULONG Signature;
+
+        /* Read the MBR */
+        if (!MachDiskReadLogicalSectors(FrldrBootDrive, 16ULL, 1, DiskReadBuffer))
+        {
+            ERR("Reading MBR failed\n");
+            return FALSE;
+        }
+
+        Buffer = (ULONG*)DiskReadBuffer;
+        Mbr = (PMASTER_BOOT_RECORD)DiskReadBuffer;
+
+        Signature = Mbr->Signature;
+        TRACE("Signature: %x\n", Signature);
+
+        /* Calculate the MBR checksum */
+        for (i = 0; i < 2048 / sizeof(ULONG); i++)
+        {
+            Checksum += Buffer[i];
+        }
+        Checksum = ~Checksum + 1;
+        TRACE("Checksum: %x\n", Checksum);
+
+        /* Fill out the ARC disk block */
+        AddReactOSArcDiskInfo(FrLdrBootPath, Signature, Checksum, TRUE);
+
+        FsRegisterDevice(FrLdrBootPath, &UefiDiskVtbl);
+        PcBiosDiskCount++; // This is not accounted for in the number of pre-enumerated BIOS drives!
+        TRACE("Additional boot drive detected: 0x%02X\n", (int)FrldrBootDrive);
+    }
+    return TRUE;
+}
+
+UCHAR
+UefiGetFloppyCount(VOID)
+{
+    /* no floppy for you */
+    return 0;
+}
+
+BOOLEAN
+UefiDiskReadLogicalSectors(
+    IN UCHAR DriveNumber,
+    IN ULONGLONG SectorNumber,
+    IN ULONG SectorCount,
+    OUT PVOID Buffer)
+{
+    PVOID mbr[2048];
+    ULONG UefiDriveNumber;
+    UefiDriveNumber =  InternalUefiDisk[DriveNumber - FIRST_BIOS_DISK].UefiRootNumber;
+    TRACE("UefiDiskReadLogicalSectors: DriveNumber: %d\n", UefiDriveNumber);
+    GlobalSystemTable->BootServices->HandleProtocol(handles[UefiDriveNumber], &bioGuid, (void **) &bio);
+
+    /* Devices setup */
+    bio->ReadBlocks(bio, bio->Media->MediaId, SectorNumber, (bio->Media->BlockSize * SectorCount),  &mbr);
+    RtlCopyMemory(Buffer, mbr, SectorCount * bio->Media->BlockSize);
```

Comment:
It seems to me that (because here we don't need to have a separate buffer as in BIOS based code) you can skip that mbr local variable directly and just use the `Buffer` variable, in the ReadBlocks call... since you read exactly the very same number of bytes.